plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,376 @@
+                                        .name(queryCreatedEvent.getMetadata().getQueryId())
+                                        .facets(
+                                                openLineage.newJobFacetsBuilder()
+                                                        .sql(openLineage.newSQLJobFacet(queryCreatedEvent.getMetadata().getQuery()))
+                                                        .build())
+                                        .build())
+                        .build();
+
+        return startEvent;
+    }
+
+    public OpenLineage.RunEvent getCompletedEvent(UUID runID, QueryCompletedEvent queryCompletedEvent)
+            throws IllegalAccessException
+    {
+        boolean failed = queryCompletedEvent.getMetadata().getQueryState().equals("FAILED");
+
+        OpenLineage.RunFacetsBuilder runFacetsBuilder = openLineage.newRunFacetsBuilder();
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_METADATA.getText(),
+                getTrinoMetadataFacet(queryCompletedEvent.getMetadata()));
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_CONTEXT.getText(),
+                getTrinoQueryContextFacet(queryCompletedEvent.getContext()));
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_STATISTICS.getText(),
+                getTrinoQueryStatisticsFacet(queryCompletedEvent.getStatistics()));
+
+        OpenLineage.RunEvent completedEvent =
+                openLineage.newRunEventBuilder()
+                        .eventType(
+                                failed
+                                        ? OpenLineage.RunEvent.EventType.FAIL
+                                        : OpenLineage.RunEvent.EventType.COMPLETE)
+                        .eventTime(queryCompletedEvent.getEndTime().atZone(ZoneId.of("UTC")))
+                        .run(openLineage.newRunBuilder().runId(runID).facets(runFacetsBuilder.build()).build())
+                        .job(
+                                openLineage.newJobBuilder()
+                                        .namespace(this.jobNamespace)
+                                        .name(queryCompletedEvent.getMetadata().getQueryId())
+                                        .facets(
+                                                openLineage.newJobFacetsBuilder()
+                                                        .sql(openLineage.newSQLJobFacet(queryCompletedEvent.getMetadata().getQuery()))
+                                                        .build())
+                                        .build())
+                        .inputs(buildInputs(queryCompletedEvent.getIoMetadata()))
+                        .outputs(buildOutputs(queryCompletedEvent.getIoMetadata()))
+                        .build();
+
+        return completedEvent;
+    }
+
+    private List<OpenLineage.InputDataset> buildInputs(QueryIOMetadata ioMetadata)
+    {
+        return ioMetadata.getInputs().stream().map(inputMetadata ->
```

Comment:
For determining can we use `queryCompletedEvent.getMetadata().getTables()` which provide additional information on whether is a table is referenced directly or via a View this could provide a better lineage information.

Reply:
I checked it and don't see much benefit over QueryIOMetadata. However, I could use your knowledge here - is there a way to get info about underlying system related to given table/view? For example, we can do `inputMetadata.getCatalogName()` but can we get some more information about this catalog? Like whether it's druid, hive, etc ?

Reply:
`queryCompletedEvent.getMetadata().getTables()#isDirectlyReferenced` provides an information if a table is referenced directly or is it referred via a view in this case we could point lineage only to the views instead of mapping them to the underlying table as well. Plus it also captures some additional information like viewText and referencedChain - So I guess for input details it makes much sense to use `queryCompletedEvent.getMetadata().getTables()`

Reply:
> Like whether it's druid, hive, etc ?

I don't think we have an option to fetch those information today - But it could be possible